src/CoreWCF.Primitives/src/CoreWCF/Description/WsdlHelper.cs

```diff
@@ -0,0 +1,328 @@
+            foreach (WsdlNS.ServiceDescription wsdl in wsdls)
+            {
+                newWsdls.Add(CloneWsdl(wsdl));
+            }
+
+            return newWsdls;
+        }
+
+        private static void MergeWsdl(WsdlNS.ServiceDescription singleWsdl, WsdlNS.ServiceDescription wsdl, Dictionary<XmlQualifiedName, XmlQualifiedName> bindingReferenceChanges)
+        {
+            if (wsdl.Services.Count > 0)
+            {
+                singleWsdl.Name = wsdl.Name;
+            }
+
+            foreach (WsdlNS.Binding binding in wsdl.Bindings)
+            {
+                string uniqueBindingName = NamingHelper.GetUniqueName(binding.Name, IsBindingNameUsed, singleWsdl.Bindings);
+                if (binding.Name != uniqueBindingName)
+                {
+                    bindingReferenceChanges.Add(
+                        new XmlQualifiedName(binding.Name, binding.ServiceDescription.TargetNamespace),
+                        new XmlQualifiedName(uniqueBindingName, singleWsdl.TargetNamespace));
+                    UpdatePolicyKeys(binding, uniqueBindingName, wsdl);
+                    binding.Name = uniqueBindingName;
+                }
+
+                singleWsdl.Bindings.Add(binding);
+            }
+
+            foreach (object extension in wsdl.Extensions)
+            {
+                singleWsdl.Extensions.Add(extension);
+            }
+
+            foreach (WsdlNS.Message message in wsdl.Messages)
+            {
+                singleWsdl.Messages.Add(message);
+            }
+
+            foreach (WsdlNS.Service service in wsdl.Services)
+            {
+                singleWsdl.Services.Add(service);
+            }
+
+            foreach (string warning in wsdl.ValidationWarnings)
+            {
+                singleWsdl.ValidationWarnings.Add(warning);
+            }
+        }
+
+        private static void UpdatePolicyKeys(WsdlNS.Binding binding, string newBindingName, WsdlNS.ServiceDescription wsdl)
+        {
+            string oldBindingName = binding.Name;
+
+            // policy
+            IEnumerable<XmlElement> bindingPolicies = FindAllElements(wsdl.Extensions, MetadataStrings.WSPolicy.Elements.Policy);
+            string policyIdStringPrefixFormat = "{0}_";
+            foreach (XmlElement policyElement in bindingPolicies)
+            {
+                XmlNode policyId = policyElement.Attributes.GetNamedItem(MetadataStrings.Wsu.Attributes.Id, MetadataStrings.Wsu.NamespaceUri);
+                string policyIdString = policyId.Value;
+                string policyIdStringWithOldBindingName = string.Format(CultureInfo.InvariantCulture, policyIdStringPrefixFormat, oldBindingName);
+                string policyIdStringWithNewBindingName = string.Format(CultureInfo.InvariantCulture, policyIdStringPrefixFormat, newBindingName);
+                if (policyId != null && policyIdString != null && policyIdString.StartsWith(policyIdStringWithOldBindingName, StringComparison.Ordinal))
```

Comment:
`policyId` cannot be null here due to `string policyIdString = policyId.Value;` on line 158

Reply:
policyId could be null in which case line 158 would throw NRE and this code wouldn't be reached. All binding polices created by CoreWCF will have an Id, but this is extensible and someone else's code might not put the Id there. The code needs to be more defensive. I think using null coalescing on line 158 is sufficient. This way line 158 can't throw a NRE and this line becomes reachable even if policyId is null, in which case the null check here becomes useful.